pkg/gatewayserver/io/basicstationlns/messages/upstream.go

```diff
@@ -0,0 +1,294 @@
+	GpsTime int64            `json:"gpstime"`
+}
+
+// MarshalJSON implements json.Marshaler.
+func (conf TxConfirmation) MarshalJSON() ([]byte, error) {
+	type Alias TxConfirmation
+	return json.Marshal(struct {
+		Type string `json:"msgtype"`
+		Alias
+	}{
+		Type:  TypeUpstreamTxConfirmation,
+		Alias: Alias(conf),
+	})
+}
+
+// ToUplinkMessage extracts fields from the basic station Join Request "jreq" message and converts them into an UplinkMessage for the network server.
+func (req *JoinRequest) ToUplinkMessage(ids ttnpb.GatewayIdentifiers, bandID string) (ttnpb.UplinkMessage, error) {
+	up := ttnpb.UplinkMessage{}
+	up.ReceivedAt = time.Now()
+
+	parsedMHDR := ttnpb.MHDR{}
+	err := lorawan.UnmarshalMHDR([]byte{byte(req.MHdr)}, &parsedMHDR)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errJoinRequestMessage.WithCause(err)
+	}
+
+	micBytes, err := getInt32AsByteSlice(req.MIC)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errJoinRequestMessage.WithCause(err)
+	}
+	up.Payload = &ttnpb.Message{
+		MHDR: parsedMHDR,
+		MIC:  micBytes,
+		Payload: &ttnpb.Message_JoinRequestPayload{JoinRequestPayload: &ttnpb.JoinRequestPayload{
+			JoinEUI:  req.JoinEUI.EUI64,
+			DevEUI:   req.DevEUI.EUI64,
+			DevNonce: [2]byte{byte(req.DevNonce), byte(req.DevNonce >> 8)},
+		}},
+	}
+
+	up.RawPayload, err = lorawan.MarshalMessage(*up.Payload)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errJoinRequestMessage.WithCause(err)
+	}
+
+	timestamp := uint32(req.RadioMetaData.UpInfo.XTime & 0xFFFFFFFF)
+
+	ulToken := ttnpb.UplinkToken{
+		GatewayAntennaIdentifiers: ttnpb.GatewayAntennaIdentifiers{
+			GatewayIdentifiers: ids,
+			AntennaIndex:       uint32(req.RadioMetaData.UpInfo.RCtx),
+		},
+		Timestamp: timestamp,
+	}
+	ulTokenBytes, err := ulToken.Marshal()
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errJoinRequestMessage.WithCause(err)
+	}
+
+	rxTime := time.Unix(req.RadioMetaData.UpInfo.RxTime, 0)
+	rxMetadata := &ttnpb.RxMetadata{
+		GatewayIdentifiers: ids,
+		Time:               &rxTime,
+		Timestamp:          timestamp,
+		RSSI:               req.RadioMetaData.UpInfo.RSSI,
+		SNR:                req.RadioMetaData.UpInfo.SNR,
+		UplinkToken:        ulTokenBytes,
+	}
+	up.RxMetadata = append(up.RxMetadata, rxMetadata)
+
+	loraDR, err := getDataRateFromIndex(bandID, req.RadioMetaData.DataRate)
+	if err != nil {
+		return ttnpb.UplinkMessage{}, errJoinRequestMessage.WithCause(err)
+	}
+	up.Settings = ttnpb.TxSettings{
+		Frequency:     req.RadioMetaData.Frequency,
+		DataRateIndex: ttnpb.DataRateIndex(req.RadioMetaData.DataRate),
+		DataRate:      loraDR,
+	}
+
+	return up, nil
+}
+
+// ToUplinkMessage extracts fields from the basic station Join Request "jreq" message and converts them into an UplinkMessage for the network server.
+func (updf *UplinkDataFrame) ToUplinkMessage(ids ttnpb.GatewayIdentifiers, bandID string) (ttnpb.UplinkMessage, error) {
+	up := ttnpb.UplinkMessage{}
+	up.ReceivedAt = time.Now()
```

Comment:
I think this is done by `pkg/gatewayserver/io`?

Reply:
`io` expects this to be filled out by the front-end and is used `as-is` in `HandleUp`.

Reply:
See above. Determine this time as soon as a message comes in, before you know a type.